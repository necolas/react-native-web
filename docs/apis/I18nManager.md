# I18nManager

Control and set the layout and writing direction of the application. You must
set `dir="rtl"` (and should set `lang="${lang}"`) on the root element of your
app.

## Properties

**isRTL**: bool = false

Whether the application is currently in RTL mode.

## Methods

static **allowRTL**(allowRTL: bool)

Allow the application to display in RTL mode.

static **forceRTL**(allowRTL: bool)

Force the application to display in RTL mode.

static **setRTL**(allowRTL: bool)

Set the application to display in RTL mode. You will need to determine the
user's preferred locale and if it is an RTL language. (This is best done on the
server as it is notoriously inaccurate to deduce client-side.)
