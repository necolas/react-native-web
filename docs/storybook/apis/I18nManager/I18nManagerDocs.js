/* eslint-disable react/jsx-sort-props */

/**
 * @flow
 */

import RTLToggleExample from './examples/RTLToggle';
import React from 'react';
import { storiesOf } from '@kadira/storybook';
import UIExplorer, { DocItem } from '../../ui-explorer';

const sections = [
  {
    title: 'Properties',
    entries: [
      <DocItem
        name="isRTL"
        typeInfo="boolean = false"
        description="Whether the application is currently in RTL mode."
      />
    ]
  },
  {
    title: 'Methods',
    entries: [
      <DocItem
        name="static allowRTL"
        typeInfo="(allowRTL: boolean) => void"
        description="Allow the application to display in RTL mode."
      />,

      <DocItem
        name="static forceRTL"
        typeInfo="(forceRTL: boolean) => void"
        description="Force the application to display in RTL mode."
      />,

      <DocItem
        label="web"
        name="static setPreferredLanguageRTL"
        typeInfo="(isRTL: boolean) => void"
        description="Set the application's preferred writing direction to RTL. You will need to determine the user's preferred locale server-side (from HTTP headers) and decide whether it's an RTL language."
      />
    ]
  },
  {
    title: 'Examples',
    entries: [
      <DocItem
        description="Toggling LTR/RTL layout at runtime"
        example={{
          render: () => <RTLToggleExample />
        }}
      />
    ]
  }
];

storiesOf('APIs', module).add('I18nManager', () =>
  <UIExplorer
    description="Control and set the layout and writing direction of the application."
    sections={sections}
    title="I18nManager"
    url="apis/I18nManager"
  />
);
