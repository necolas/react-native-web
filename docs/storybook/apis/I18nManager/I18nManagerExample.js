import { storiesOf } from '@kadira/storybook';
import UIExplorer from '../../UIExplorer';
import { I18nManager, StyleSheet, TouchableHighlight, Text, View } from 'react-native';
import React, { Component } from 'react';

class I18nManagerExample extends Component {
  componentWillUnmount() {
    I18nManager.setPreferredLanguageRTL(false);
  }

  render() {
    return (
      <View style={styles.container}>
        <Text accessibilityRole="heading" style={styles.welcome}>
          LTR/RTL layout example!
        </Text>
        <Text style={styles.text}>
          The writing direction of text is automatically determined by the browser, independent of
          the global writing direction of the app.
        </Text>
        <Text style={[styles.text, styles.rtlText]}>
          أحب اللغة العربية
        </Text>
        <Text style={[styles.text, styles.textAlign]}>
          textAlign toggles
        </Text>
        <View style={styles.horizontal}>
          <View style={[styles.box, { backgroundColor: 'lightblue' }]}>
            <Text>One</Text>
          </View>
          <View style={[styles.box]}>
            <Text>Two</Text>
          </View>
        </View>
        <TouchableHighlight
          onPress={this._handleToggle}
          style={styles.toggle}
          underlayColor="rgba(0,0,0,0.25)"
        >
          <Text>Toggle LTR/RTL</Text>
        </TouchableHighlight>
      </View>
    );
  }

  _handleToggle = () => {
    I18nManager.setPreferredLanguageRTL(!I18nManager.isRTL);
    this.forceUpdate();
  };
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#F5FCFF',
    flex: 1,
    justifyContent: 'center',
    padding: 10
  },
  welcome: {
    fontSize: 28,
    marginVertical: 10
  },
  text: {
    color: '#333333',
    fontSize: 18,
    marginBottom: 5
  },
  textAlign: {
    textAlign: 'left'
  },
  horizontal: {
    flexDirection: 'row',
    marginVertical: 10
  },
  box: {
    borderWidth: 1,
    flex: 1
  },
  toggle: {
    alignSelf: 'center',
    borderColor: 'black',
    borderStyle: 'solid',
    borderWidth: 1,
    marginTop: 10,
    padding: 10
  }
});

const examples = [
  {
    title: 'RTL toggle',
    render: () => <I18nManagerExample />
  }
];

storiesOf('APIs', module).add('I18nManager', () =>
  <UIExplorer examples={examples} title="I18nManager" />
);
