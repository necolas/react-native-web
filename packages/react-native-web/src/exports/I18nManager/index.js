/**
 * Copyright (c) Nicolas Gallagher.
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

type I18nManagerStatus = {
  allowRTL: (allowRTL: boolean) => void,
  forceRTL: (forceRTL: boolean) => void,
  getConstants: () => Constants
};

type Constants = {
  isRTL: boolean
};

const I18nManager: I18nManagerStatus = {
  allowRTL() {
    return;
  },
  forceRTL() {
    return;
  },
  getConstants(): Constants {
    return { isRTL: false };
  }
};

export default I18nManager;
