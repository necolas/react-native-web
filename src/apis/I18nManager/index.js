/**
 * Copyright (c) 2016-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @providesModule I18nManager
 * @flow
 */

import ExecutionEnvironment from 'fbjs/lib/ExecutionEnvironment';

type I18nManagerStatus = {
  allowRTL: (allowRTL: boolean) => void,
  forceRTL: (forceRTL: boolean) => void,
  setPreferredLanguageRTL: (setRTL: boolean) => void,
  isRTL: boolean
};

let isPreferredLanguageRTL = false;
let isRTLAllowed = true;
let isRTLForced = false;

const isRTL = () => {
  if (isRTLForced) {
    return true;
  }
  return isRTLAllowed && isPreferredLanguageRTL;
};

const onChange = () => {
  if (ExecutionEnvironment.canUseDOM) {
    if (document.documentElement && document.documentElement.setAttribute) {
      document.documentElement.setAttribute('dir', isRTL() ? 'rtl' : 'ltr');
    }
  }
};

const I18nManager: I18nManagerStatus = {
  allowRTL(bool) {
    isRTLAllowed = bool;
    onChange();
  },
  forceRTL(bool) {
    isRTLForced = bool;
    onChange();
  },
  setPreferredLanguageRTL(bool) {
    isPreferredLanguageRTL = bool;
    onChange();
  },
  get isRTL() {
    return isRTL();
  }
};

export default I18nManager;
