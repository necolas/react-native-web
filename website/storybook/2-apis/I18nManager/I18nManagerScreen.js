/* eslint-disable react/jsx-sort-props */

/**
 * @flow
 */

import RTLToggle from './examples/RTLToggle';
import React from 'react';
import UIExplorer, { Description, DocItem, Section, storiesOf } from '../../ui-explorer';

const I18nManagerScreen = () => (
  <UIExplorer title="I18nManager" url="2-apis/I18nManager">
    <Description>
      Control and query the layout and writing direction of the application.
    </Description>
    <Section title="Properties">
      <DocItem
        name="isRTL"
        typeInfo="boolean = false"
        description="Whether the application is currently in RTL mode."
      />

      <DocItem
        name="doLeftAndRightSwapInRTL"
        typeInfo="boolean = true"
        description="Whether the application swaps left/right styles in RTL mode."
      />
    </Section>

    <Section title="Methods">
      <DocItem
        name="allowRTL"
        typeInfo="(allowRTL: boolean) => void"
        description="Allow the application to display in RTL mode."
      />

      <DocItem
        name="forceRTL"
        typeInfo="(forceRTL: boolean) => void"
        description="Force the application to display in RTL mode."
      />

      <DocItem
        name="swapLeftAndRightInRTL"
        typeInfo="(flipStyles: boolean) => void"
        description="Control whether the application swaps left/right styles in RTL mode. Applications relying on start/end styles may prefer to disable automatic BiDi-flipping of left/right styles."
      />

      <DocItem
        label="web"
        name="setPreferredLanguageRTL"
        typeInfo="(isRTL: boolean) => void"
        description="Set the application's preferred writing direction to RTL. You may need to infer the user's preferred locale on the server (from HTTP headers) and decide whether it's an RTL language."
      />
    </Section>

    <Section title="Examples">
      <DocItem
        description="Toggling LTR/RTL layout at runtime"
        example={{
          render: () => <RTLToggle />
        }}
      />
    </Section>
  </UIExplorer>
);

storiesOf('APIs', module).add('I18nManager', I18nManagerScreen);
