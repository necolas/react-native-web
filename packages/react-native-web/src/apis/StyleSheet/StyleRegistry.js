/**
 * Copyright (c) 2016-present, Nicolas Gallagher.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * WARNING: changes to this file in particular can cause significant changes to
 * the results of render performance benchmarks.
 *
 * @noflow
 */

import createReactDOMStyle from './createReactDOMStyle';
import flattenArray from '../../modules/flattenArray';
import flattenStyle from './flattenStyle';
import I18nManager from '../I18nManager';
import i18nStyle from './i18nStyle';
import { prefixInlineStyles } from '../../modules/prefixStyles';
import ReactNativePropRegistry from '../../modules/ReactNativePropRegistry';
import StyleManager from './StyleManager';

const emptyObject = {};

const createCacheKey = id => {
  const prefix = 'rn';
  return `${prefix}-${id}`;
};

const classListToString = list => list.join(' ').trim();

export default class StyleRegistry {
  cache = { ltr: {}, rtl: {} };
  styleManager = new StyleManager();

  getStyleSheets() {
    return this.styleManager.getStyleSheets();
  }

  /**
   * Registers and precaches a React Native style object to HTML class names
   */
  register(flatStyle) {
    const id = ReactNativePropRegistry.register(flatStyle);
    this._registerById(id);
    return id;
  }

  _registerById(id) {
    const dir = I18nManager.isRTL ? 'rtl' : 'ltr';
    if (!this.cache[dir][id]) {
      const style = flattenStyle(id);
      const domStyle = createReactDOMStyle(i18nStyle(style));
      Object.keys(domStyle).forEach(styleProp => {
        const value = domStyle[styleProp];
        if (value != null) {
          this.styleManager.setDeclaration(styleProp, value);
        }
      });
      this.cache[dir][id] = true;
    }
  }

  /**
   * Resolves a React Native style object to DOM attributes
   */
  resolve(reactNativeStyle, options = emptyObject) {
    if (!reactNativeStyle) {
      return emptyObject;
    }

    // fast and cachable
    if (typeof reactNativeStyle === 'number') {
      this._registerById(reactNativeStyle);
      const key = createCacheKey(reactNativeStyle);
      return this._resolveStyleIfNeeded(reactNativeStyle, options, key);
    }

    // resolve a plain RN style object
    if (!Array.isArray(reactNativeStyle)) {
      return this._resolveStyle(reactNativeStyle, options);
    }

    // flatten the style array
    // cache resolved props when all styles are registered
    // otherwise fallback to resolving
    const flatArray = flattenArray(reactNativeStyle);
    let isArrayOfNumbers = true;
    for (let i = 0; i < flatArray.length; i++) {
      const id = flatArray[i];
      if (typeof id !== 'number') {
        isArrayOfNumbers = false;
      } else {
        this._registerById(id);
      }
    }
    const key = isArrayOfNumbers ? createCacheKey(flatArray.join('-')) : null;
    return this._resolveStyleIfNeeded(flatArray, options, key);
  }

  /**
   * Resolves a React Native style object to DOM attributes, accounting for
   * the existing styles applied to the DOM node.
   *
   * To determine the next style, some of the existing DOM state must be
   * converted back into React Native styles.
   */
  resolveStateful(rnStyleNext, domStyleProps, options) {
    const { classList: rdomClassList, style: rdomStyle } = domStyleProps;

    // Convert the DOM classList back into a React Native form
    // Preserves unrecognized class names.
    const { classList: rnClassList, style: rnStyle } = rdomClassList.reduce(
      (styleProps, className) => {
        const { prop, value } = this.styleManager.getDeclaration(className);
        if (prop) {
          styleProps.style[prop] = value;
        } else {
          styleProps.classList.push(className);
        }
        return styleProps;
      },
      { classList: [], style: {} }
    );

    // Create next DOM style props from current and next RN styles
    const { classList: rdomClassListNext, style: rdomStyleNext } = this.resolve(
      [rnStyle, rnStyleNext],
      options
    );

    // Next class names take priority over current inline styles
    const style = { ...rdomStyle };
    rdomClassListNext.forEach(className => {
      const { prop } = this.styleManager.getDeclaration(className);
      if (style[prop]) {
        style[prop] = '';
      }
    });

    // Next inline styles take priority over current inline styles
    Object.assign(style, rdomStyleNext);

    // Add the current class names not managed by React Native
    const className = classListToString(rdomClassListNext.concat(rnClassList));

    return { className, style };
  }

  /**
   * Resolves a React Native style object
   */
  _resolveStyle(reactNativeStyle, options) {
    const flatStyle = flattenStyle(reactNativeStyle);
    const domStyle = createReactDOMStyle(options.i18n === false ? flatStyle : i18nStyle(flatStyle));

    const props = Object.keys(domStyle).reduce(
      (props, styleProp) => {
        const value = domStyle[styleProp];
        if (value != null) {
          const className = this.styleManager.getClassName(styleProp, value);
          if (className) {
            props.classList.push(className);
          } else {
            if (!props.style) {
              props.style = {};
            }
            // 4x slower render
            props.style[styleProp] = value;
          }
        }
        return props;
      },
      { classList: [] }
    );

    props.className = classListToString(props.classList);
    if (props.style) {
      props.style = prefixInlineStyles(props.style);
    }
    return props;
  }

  /**
   * Caching layer over 'resolveStyle'
   */
  _resolveStyleIfNeeded(style, options, key) {
    const dir = I18nManager.isRTL ? 'rtl' : 'ltr';
    if (key) {
      if (!this.cache[dir][key]) {
        // slow: convert style object to props and cache
        this.cache[dir][key] = this._resolveStyle(style, options);
      }
      return this.cache[dir][key];
    }
    return this._resolveStyle(style, options);
  }
}
