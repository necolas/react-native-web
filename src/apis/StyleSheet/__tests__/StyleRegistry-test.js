/* eslint-env jasmine, jest */

import I18nManager from '../../I18nManager';
import StyleRegistry from '../StyleRegistry';

let styleRegistry;

describe('apis/StyleSheet/StyleRegistry', () => {
  beforeEach(() => {
    styleRegistry = new StyleRegistry();
  });

  test('register', () => {
    const style = { opacity: 0 };
    const id = styleRegistry.register(style);
    expect(typeof id === 'number').toBe(true);
  });

  describe('resolve', () => {
    const styleA = { borderWidth: 0, borderColor: 'red', width: 100 };
    const styleB = {
      position: 'absolute',
      left: 50,
      pointerEvents: 'box-only'
    };
    const styleC = { width: 200 };

    const testResolve = (a, b, c) => {
      // no common properties, different resolving order, same result
      const resolve1 = styleRegistry.resolve([a, b]);
      expect(resolve1).toMatchSnapshot();
      const resolve2 = styleRegistry.resolve([b, a]);
      expect(resolve1).toEqual(resolve2);

      // common properties, different resolving order, different result
      const resolve3 = styleRegistry.resolve([a, b, c]);
      expect(resolve3).toMatchSnapshot();
      const resolve4 = styleRegistry.resolve([c, a, b]);
      expect(resolve4).toMatchSnapshot();
      expect(resolve3).not.toEqual(resolve4);
    };

    test('with register, resolves to className', () => {
      const a = styleRegistry.register(styleA);
      const b = styleRegistry.register(styleB);
      const c = styleRegistry.register(styleC);
      testResolve(a, b, c);
    });

    test('with register before RTL, resolves to className', () => {
      const a = styleRegistry.register({ left: '12.34%' });
      const b = styleRegistry.register({ textAlign: 'left' });
      const c = styleRegistry.register({ marginLeft: 10 });
      I18nManager.forceRTL(true);
      const resolved = styleRegistry.resolve([a, b, c]);
      I18nManager.forceRTL(false);
      expect(resolved).toMatchSnapshot();
    });

    test('with register, resolves to mixed', () => {
      const a = styleA;
      const b = styleRegistry.register(styleB);
      const c = styleRegistry.register(styleC);
      testResolve(a, b, c);
    });

    test('without register, resolves to inline styles', () => {
      testResolve(styleA, styleB, styleC);
    });
  });

  describe('resolveStateful', () => {
    test('preserves unrelated class names', () => {
      const domStyleProps = { classList: ['unknown-class-1', 'unknown-class-2'], style: {} };
      const domStyleNextProps = styleRegistry.resolveStateful({}, domStyleProps);
      expect(domStyleNextProps).toMatchSnapshot();
    });

    test('preserves unrelated inline styles', () => {
      const domStyleProps = { classList: [], style: { fontSize: '20px' } };
      const domStyleNextProps = styleRegistry.resolveStateful({ opacity: 1 }, domStyleProps);
      expect(domStyleNextProps).toMatchSnapshot();
    });

    test('next class names have priority over current inline styles', () => {
      const domStyleProps = { classList: [], style: { opacity: 0.5 } };
      const nextStyle = styleRegistry.register({ opacity: 1 });
      const domStyleNextProps = styleRegistry.resolveStateful(nextStyle, domStyleProps);
      expect(domStyleNextProps).toMatchSnapshot();
    });

    test('next inline styles have priority over current inline styles', () => {
      // note: this also checks for correctly uppercasing the first letter of DOM vendor prefixes
      const domStyleProps = {
        classList: [],
        style: { opacity: 0.5, WebkitTransform: 'scale(1)', transform: 'scale(1)' }
      };
      const domStyleNextProps = styleRegistry.resolveStateful(
        { opacity: 1, transform: [{ scale: 2 }] },
        domStyleProps
      );
      expect(domStyleNextProps).toMatchSnapshot();
    });
  });
});
