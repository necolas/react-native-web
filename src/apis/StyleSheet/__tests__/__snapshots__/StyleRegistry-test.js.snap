// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apis/StyleSheet/StyleRegistry resolve with register before RTL, resolves to className 1`] = `
Object {
  "classList": Array [
    "rn-marginRight-zso239",
    "rn-right-1bnbe1j",
    "rn-textAlign-1ff274t",
  ],
  "className": "rn-marginRight-zso239 rn-right-1bnbe1j rn-textAlign-1ff274t",
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to className 1`] = `
Object {
  "classList": Array [
    "rn-borderTopColor-1gxhl28",
    "rn-borderRightColor-knoah9",
    "rn-borderBottomColor-1ani3fp",
    "rn-borderLeftColor-ribj9x",
    "rn-borderTopWidth-13yce4e",
    "rn-borderRightWidth-fnigne",
    "rn-borderBottomWidth-ndvcnb",
    "rn-borderLeftWidth-gxnn5r",
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
    "rn-width-b8lwoo",
  ],
  "className": "rn-borderTopColor-1gxhl28 rn-borderRightColor-knoah9 rn-borderBottomColor-1ani3fp rn-borderLeftColor-ribj9x rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d rn-width-b8lwoo",
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to className 2`] = `
Object {
  "classList": Array [
    "rn-borderTopColor-1gxhl28",
    "rn-borderRightColor-knoah9",
    "rn-borderBottomColor-1ani3fp",
    "rn-borderLeftColor-ribj9x",
    "rn-borderTopWidth-13yce4e",
    "rn-borderRightWidth-fnigne",
    "rn-borderBottomWidth-ndvcnb",
    "rn-borderLeftWidth-gxnn5r",
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
    "rn-width-l0gwng",
  ],
  "className": "rn-borderTopColor-1gxhl28 rn-borderRightColor-knoah9 rn-borderBottomColor-1ani3fp rn-borderLeftColor-ribj9x rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d rn-width-l0gwng",
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to className 3`] = `
Object {
  "classList": Array [
    "rn-borderTopColor-1gxhl28",
    "rn-borderRightColor-knoah9",
    "rn-borderBottomColor-1ani3fp",
    "rn-borderLeftColor-ribj9x",
    "rn-borderTopWidth-13yce4e",
    "rn-borderRightWidth-fnigne",
    "rn-borderBottomWidth-ndvcnb",
    "rn-borderLeftWidth-gxnn5r",
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
    "rn-width-b8lwoo",
  ],
  "className": "rn-borderTopColor-1gxhl28 rn-borderRightColor-knoah9 rn-borderBottomColor-1ani3fp rn-borderLeftColor-ribj9x rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d rn-width-b8lwoo",
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to mixed 1`] = `
Object {
  "classList": Array [
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
  ],
  "className": "rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "width": "100px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to mixed 2`] = `
Object {
  "classList": Array [
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
    "rn-width-l0gwng",
  ],
  "className": "rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d rn-width-l0gwng",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve with register, resolves to mixed 3`] = `
Object {
  "classList": Array [
    "rn-left-1tsx3h3",
    "rn-pointerEvents-ah5dr5",
    "rn-position-u8s1d",
  ],
  "className": "rn-left-1tsx3h3 rn-pointerEvents-ah5dr5 rn-position-u8s1d",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "width": "100px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve without register, resolves to inline styles 1`] = `
Object {
  "classList": Array [
    "rn-pointerEvents-ah5dr5",
  ],
  "className": "rn-pointerEvents-ah5dr5",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "left": "50px",
    "position": "absolute",
    "width": "100px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve without register, resolves to inline styles 2`] = `
Object {
  "classList": Array [
    "rn-pointerEvents-ah5dr5",
  ],
  "className": "rn-pointerEvents-ah5dr5",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "left": "50px",
    "position": "absolute",
    "width": "200px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolve without register, resolves to inline styles 3`] = `
Object {
  "classList": Array [
    "rn-pointerEvents-ah5dr5",
  ],
  "className": "rn-pointerEvents-ah5dr5",
  "style": Object {
    "borderBottomColor": "red",
    "borderBottomWidth": "0px",
    "borderLeftColor": "red",
    "borderLeftWidth": "0px",
    "borderRightColor": "red",
    "borderRightWidth": "0px",
    "borderTopColor": "red",
    "borderTopWidth": "0px",
    "left": "50px",
    "position": "absolute",
    "width": "100px",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolveStateful next class names have priority over current inline styles 1`] = `
Object {
  "className": "rn-opacity-6dt33c",
  "style": Object {
    "opacity": "",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolveStateful next inline styles have priority over current inline styles 1`] = `
Object {
  "className": "",
  "style": Object {
    "WebkitTransform": "scale(2)",
    "opacity": 1,
    "transform": "scale(2)",
  },
}
`;

exports[`apis/StyleSheet/StyleRegistry resolveStateful preserves unrelated class names 1`] = `
Object {
  "className": "unknown-class-1 unknown-class-2",
  "style": Object {},
}
`;

exports[`apis/StyleSheet/StyleRegistry resolveStateful preserves unrelated inline styles 1`] = `
Object {
  "className": "",
  "style": Object {
    "fontSize": "20px",
    "opacity": 1,
  },
}
`;
